fakeredis/_server.py

```diff
@@ -1897,19 +1898,58 @@ def _apply_withscores(items, withscores):
 
     @command((Key(ZSet), bytes, bytes), (bytes,))
     def zadd(self, key, *args):
-        # TODO: handle NX, XX, CH, INCR
+        # TODO: handle INCR
         zset = key.value
-        if len(args) % 2 != 0:
+
+        i = 0
+        ch = False
+        nx = False
+        xx = False
+        while i < len(args):
+            if casematch(args[i], b'ch'):
+                ch = True
+                i += 1
+            elif casematch(args[i], b'nx'):
+                nx = True
+                i += 1
+            elif casematch(args[i], b'xx'):
+                xx = True
+                i += 1
+            else:
+                # First argument not matching flags indicates the start of
+                # score pairs.
+                break
+
+        if nx and xx:
+            raise redis.ResponseError(ZADD_NX_XX_ERROR_MSG)
+
+        elements = args[i:]
+        if not elements or len(elements) % 2 != 0:
             raise redis.ResponseError(SYNTAX_ERROR_MSG)
-        items = []
         # Parse all scores first, before updating
-        for i in range(0, len(args), 2):
-            score = Float.decode(args[i])
-            items.append((score, args[i + 1]))
+        items = [
+            (Float.decode(elements[j]), elements[j + 1])
+            for j in range(0, len(elements), 2)
+        ]
         old_len = len(zset)
-        for item in items:
-            if zset.add(item[1], item[0]):
-                key.updated()
+        changed_items = 0
+        # Cache value to use on each iteration
+        not_nx_nor_xx = not nx and not xx
+
+        for item_score, item_name in items:
+            if (
+                not_nx_nor_xx
+                or (nx and item_name not in zset)
+                or (xx and item_name in zset)
+            ):
```

Comment:
I think this condition can be simplified to `if (not nx or item_name not in zset) and (not xx or item_name in zset)`.

Reply:
Fixed!